cmd/dvol-docker-plugin/dvol-docker-plugin.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+    "fmt"
+    "encoding/json"
+    "errors"
+    "log"
+    "net"
+    "net/http"
+    "os"
+    "io/ioutil"
+
+	"github.com/ClusterHQ/dvol/pkg/api"
+)
+
+const PLUGINS_DIR = "/run/docker/plugins"
+const DVOL_SOCKET = PLUGINS_DIR + "/dvol.sock"
+const VOL_DIR = "/var/lib/dvol/volumes"
+
+type ResponseImplements struct {
+    // A response to the Plugin.Activate request
+    Implements []string
+}
+
+type RequestCreate struct {
+    // A request to create a volume for Docker
+    Name string
+    Opts map[string]string
+}
+
+type RequestMount struct {
+    // A request to mount a volume for Docker
+    Name string
+}
+
+type RequestRemove struct {
+    // A request to remove a volume for Docker
+    Name string
+}
+
+type ResponseSimple struct {
+    // A response which only indicates if there was an error or not
+    Err string
+}
+
+type ResponseMount struct {
+    // A response to the VolumeDriver.Mount request
+    Mountpoint string
+    Err string
+}
+
+func main () {
+    if _, err := os.Stat(PLUGINS_DIR); err != nil {
+        if err := os.MkdirAll(PLUGINS_DIR, 0700); err != nil {
+            log.Fatalf("Could not make plugin directory %s: %v", PLUGINS_DIR, err)
+        }
+    }
+    if _, err := os.Stat(DVOL_SOCKET); err == nil {
+        if err = os.Remove(DVOL_SOCKET); err != nil {
+            log.Fatalf("Could not clean up existing socket at %s: %v", DVOL_SOCKET, err)
+        }
+    }
+    if _, err := os.Stat(VOL_DIR); err != nil {
+        if err := os.MkdirAll(VOL_DIR, 0700); err != nil {
+            log.Fatalf("Could not make volumes directory %s: %v", VOL_DIR, err)
+        }
+    }
+
+    listener, err := net.Listen("unix", DVOL_SOCKET)
```

Comment:
Kind of unfortunate that listener ends up being used so far from where it is set (at the end of the function), but it looks like you might be constrained by the API. Possible way of alleviating this is to wrap some of the intervening code in a function, like setupHttpHandlers or somesuch